### ISSUE TITLE
ModelBinder doesn't work with JsonPropertyName with FromQuery model

### ISSUE THREAD

**Issue by CaCTuCaTu4ECKuu** [dotnet/aspnetcore#27489]
Hi there. Sound complicated, but basically:
I got model with property
```
MyModel : IRequestWhatever
{
   [JsonPropertyName("some-name")]
   [JsonProperty("some-name")]
   public string MyProp {get;set;}
}
```
I got controller with action
`public IActionResult Get([FromQuery]MyModel model)`

When executing query Model Binder ignores properties with characters like "-", "_", "." (I tested)  and just leave default values...

Also, I'm on edge, I already closed so many holes, is there any workaround or I'm writing custom ModelBinder? This will not be fixed for a long time for sure...

**Comment by javiercn** [comment]
@CaCTuCaTu4ECKuu thanks for contacting us.

`JsonPropertyName` and `JsonProperty` are only used when binding JSON by the `JsonInputFormatter`. They don't apply when you use `[FromQuery]` since that uses a different binder

**Comment by CaCTuCaTu4ECKuu** [comment]
@javiercn so why then it works when i only use alphanumeric characters and mark custom property names using `JsonPropertyName `and `JsonProperty`?

**Comment by CaCTuCaTu4ECKuu** [comment]
I'm discouraged how much of essential things not possible with aspnetcore and require days of digging just to create a custom solution of 1-2 small classes which would not be needed if would be considered in first place...

If you need to use JsonProperty attribute with FromQuery or you don't want to link your domain library with MVC for that matter.
Feel free to use.
```
public class JsonPropertyQueryModelBinder : IModelBinder
{
    public Task BindModelAsync(ModelBindingContext bindingContext)
    {
        if (bindingContext == null)
            throw new ArgumentNullException(nameof(bindingContext));

        Dictionary<PropertyInfo, string> modelProps = new Dictionary<PropertyInfo, string>();
        foreach (var prop in bindingContext.ModelType.GetProperties())
        {
            var jp = prop.GetCustomAttributes<JsonPropertyNameAttribute>().FirstOrDefault();
            if (jp != null)
                modelProps.Add(prop, jp.Name);
            else
                modelProps.Add(prop, prop.Name);
        }

        Dictionary<string, string> requestProps = new Dictionary<string, string>();
        foreach (var prop in modelProps)
        {
            var val = bindingContext.ValueProvider.GetValue(prop.Value);
            //bindingContext.ModelState.SetModelValue(prop.Value, val);
            if (val != null)
                requestProps.Add(prop.Value, val.FirstValue);
        }

        var model = Activator.CreateInstance(bindingContext.ModelType);
        foreach (var prop in modelProps)
        {
            if (requestProps.ContainsKey(prop.Value))
            {
                //var modelMetadata = bindingContext.ModelMetadata.GetMetadataForType(prop.Key.PropertyType);
                var value = requestProps[prop.Value];

                var converter = TypeDescriptor.GetConverter(prop.Key.PropertyType);
                if (converter.CanConvertFrom(value.GetType()))
                {
                    prop.Key.SetValue(model, converter.ConvertFrom(value));
                    //bindingContext.ValidationState[prop.Value] = new ValidationStateEntry()
                    //{
                    //    Metadata = modelMetadata
                    //};
                }
                else
                    bindingContext.ModelState.TryAddModelError(prop.Value, $"The value \"{prop.Value}\" is not valid.");
            }
        }

        bindingContext.Result = ModelBindingResult.Success(model);
        return Task.CompletedTask;
    }
}
```

**Comment by mkArtakMSFT** [comment]
Thanks for contacting us.
@pranavkm can you please help @CaCTuCaTu4ECKuu  with their problem? Thanks!

**Comment by pranavkm** [comment]
@CaCTuCaTu4ECKuu I'm not entirely sure why `JsonProperty` \ `JsonPropertyName` would work for you some of the time. As @javiercn mentioned, `FromQuery` uses model binding which does not use System.Text.Json or Newtonsoft.Json primitives. That said, you should be able to annotate the property using `ModelBinderAttribute`. This should remove the need to have the custom model binder (I really wouldn't recommend that as a solution here):

```C#
MyModel : IRequestWhatever
{
   [ModelBinder(Name = "some-name")]
   public string MyProp {get;set;}
}
```

If you continue seeing the original issue once you've used the `ModelBinderAttribute`, we'd be happy to investigate.

**Comment by CaCTuCaTu4ECKuu** [comment]
@pranavkm yes

> I'm not entirely sure why `JsonProperty` \ `JsonPropertyName` would work for you some of the time. As @javiercn mentioned, `FromQuery` uses model binding which does not use System.Text.Json or Newtonsoft.Json primitives.
Yes, my bad, it doesnt.

ModelBinderAttribute is a part of Microsoft.AspNetCore.Mvc - why would I reference it into library when aspnetcore obviously already works with Newtonsoft.Json and System.Text.Json? And there is no (as far as I see) convinient way to map property names.
As example I'm using mediator and swagger, I distributed everything perfectly, but I have one property of my type EnumString that is technically complex type, but basically is string, I tried to use string and hide it with JsonIgnore but it's not working and I expect you know how it's mapped by default (it is shown as query property propName.Name).

I mean I have no proplems that it's not in aspnetcore or MVC itself, but those are thing people can often make use of, is there now way to add some kind of extension library for this or I dont know, add support inside, because as I said, it's inconsistent, there is support in one places and no support in another, there's bug with DateTime and modelbinder or smth I dont remember now with opened issue for few years already...
